### ISSUE TITLE
Adding embedding layer to model

### ISSUE THREAD

**Issue by laurendelong21** [babylonhealth/rgat#6]
Hello,

I'd like to include an additional dense layer to create lower dimensional feature embeddings before the RGAT layers in my model, but I continue to get variants of the following error

ValueError: Tensor("rgatn_model/dense/kernel/Read/ReadVariableOp:0", shape=(10776, 50), dtype=float32) must be from the same graph as Tensor("SparseToDense:0", shape=(5582, 10776), dtype=float32).

Below is the code for my model, which works when I do not include the embedding layer:

class RGATNModel(tf.keras.Model):
    def __init__(self, params, training):
        super(RGATNModel, self).__init__()
        self.units = params.units
        self.heads = params.heads
        self.relations = params.relations
        self.head_aggregation = params.head_aggregation
        self.attention_mode = params.attention_mode
        self.attention_style = params.attention_style
        self.attention_units = params.attention_units
        self.perms = params.perms
        self.ntypes = params.perm_types
        self.output_dim = self.units[1]

        self.embed_layer = tf.keras.layers.Dense(
                            50, activation="relu", use_bias=True,
                            kernel_initializer='glorot_uniform',
                            bias_initializer='zeros', kernel_regularizer=None,
                            bias_regularizer=None, activity_regularizer=None, kernel_constraint=None,
                            bias_constraint=None)

        self.rgat1 = RGAT(units=self.units[0],
                          relations=self.relations,
                          heads=self.heads,
                          head_aggregation=self.head_aggregation,
                          attention_mode=self.attention_mode,
                          attention_style=self.attention_style,
                          attention_units=self.attention_units,
                          activation=tf.nn.relu)
        self.rgat2 = RGAT(units=self.units[1],
                          relations=self.relations,
                          attention_mode=self.attention_mode,
                          attention_style=self.attention_style,
                          attention_units=self.attention_units)
        self.decoder = DistMult(perms=self.perms, ntypes=self.ntypes, output_dim=self.output_dim)

    def call(self, inputs, support):
        embedding = self.embed_layer(inputs=tf.sparse.to_dense(inputs))
        x = self.rgat1(inputs=embedding, support=support)
        x1 = self.rgat2(inputs=x, support=support)
        return self.decoder.call(inputs=x1)

I also noticed that it works if I add layers after the RGAT layers, but not before- do you have a suggestion to fix the issue? Thanks in advance

**Comment by laurendelong21** [comment]
Thanks for the response! Unfortunately that didn't work. It was a good thought but I am not all that sure that the error here is thrown by the mismatch in dimensionality (although maybe that error would have been thrown if I fixed this one). Here, it seems that the inputs and the learnable parameters in the new layer and not being made as part of the same computational graph, but I can not figure out why.

**Comment by franciscovargas** [comment]
You are completely right, I totally misread the message (missed the graph part). I cant quite understand why the embedding layer is detached from the computation graph :/ does it tell you which line breaks in the method call ?

**Comment by franciscovargas** [comment]
Have you tried maybe forcing everything to be on the same gaph:

```python
def call(self, inputs, support):
    g1 = tf.Graph()
    with g1.as_default():
        embedding = self.embed_layer(inputs=tf.transpose(tf.sparse.to_dense(inputs)))
        x = self.rgat1(inputs=tf.transpose(embedding), support=support)
        x1 = self.rgat2(inputs=x, support=support)
        return self.decoder.call(inputs=x1)
```
